sycl/source/device.cpp

```diff
@@ -314,17 +314,19 @@ bool device::ext_oneapi_supports_cl_extension(
       Name.data(), VersionPtr, ipVersion);
 }
 
-std::string device::ext_oneapi_cl_profile() const {
+detail::string device::ext_oneapi_cl_profile_impl() const {
```

Comment:
Unfortunately this introduces unspecified `device::ext_oneapi_cl_profile_impl ` method to the `sycl::device` class, we don't usually expose such implementation methods in sycl classes.

Please hide this implementation method into device_impl class. Please see, for example, ext_oneapi_can_compile implentation in this file:
// kernel_compiler extension methods
```
bool device::ext_oneapi_can_compile(
    ext::oneapi::experimental::source_language Language) {
  return impl->extOneapiCanCompile(Language);
}
```

Reply:
Actually, this method belongs to the private section, so this change is not publicize unspecified method to the class.
Users won't be able to use this method like all other private methods in the `sycl::device` class.

Reply:
I see, technically that works, and I see other private methods there but they were all added in scope of same abi effort, all other implementation methods are in device_impl - this is our regular approach. adding method to device_imple instead of private section also means you don't need to introduce new symbol into linux/windows dumps.
 
is there any reason why we use private methods approach just for abi related patches?

Reply:
Ok, let me approve considering that similar private methods has been already added for the same reason.

Reply:
> is there any reason why we use private methods approach just for abi related patches?

My understanding is that we can't change signature of a public method. Even if we had (or already have) some implicit conversions from `detail::string` to `std::string`, returning the former from the public method could still break someone's code. For example because they are storing the result into `their::string` also using implicit conversion from `std::string` being completely unaware of our `detail::string`.

`impl` is already within the library so even if we add `impl->right_method` there will still be `std::string` crossing ABI boundary that is part of the public API.